query/logicalplan/optimize.go

```diff
@@ -60,6 +60,25 @@ func (p *PhysicalProjectionPushDown) optimize(plan *LogicalPlan, columnsUsed []C
 type ProjectionPushDown struct{}
 
 func (p *ProjectionPushDown) Optimize(plan *LogicalPlan) *LogicalPlan {
+
+	// Don't perform the optimization if the intersection of the columns affected is less than the number of columns filtered.
+	// Otherwise we'll removed the columns we're filtering on before we filter.
+	projectColumns := projectionColumns(plan)
+	projectMap := map[string]bool{}
+	filterColumns := filterColumns(plan)
+	intersection := map[string]struct{}{}
+	for _, m := range projectColumns {
+		projectMap[m.Name()] = true
+	}
+	for _, m := range filterColumns {
+		if projectMap[m.Name()] {
+			intersection[m.Name()] = struct{}{}
+		}
+	}
+	if len(intersection) < len(filterColumns) {
+		return plan
+	}
```

Comment:
If I understand correctly, we are trying to answer the question: Is it "safe" to push down a projection in the presence of a filter?

Instead of `len(intersection) < len(filterColumns)`, could we frame the logic to be "if the set difference `set(filterColumns) - set(projectColumns) > 0` we won't optimize? 
This could allow us to simplify the logic to something like:
```go
projectColumns := projectionColumns(plan)
filterColumns := filterColumns(plan)
projectMap := map[string]struct{}{}
for _, m := range projectColumns {
    projectMap[m.Name()] = struct{}{}
}

for _, m := range filterColumns {
    if _, ok := projectMap[m.Name()]; !ok {
        // There is something that is in filter but not in
        // project => non empty set difference, short circuit
        // and exit.
        return plan
    }
}
...
```
Maybe I've misunderstood something, please lmk if that's the case!

Reply:
No you're right. I overcomplicated that. Thanks!

Reply:
I'll hold off on reviewing until after this is resolved then.